shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -37,52 +37,66 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
     val (tparams: List[TypeName], targs: List[TypeVariable]) = tparamsargs.unzip
-    def thisTy(implicit prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
+    def thisTy(prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
     def wrapMethod(pt: PolymorphicType, prov: TypeProvenance): MethodType =
-      MethodType(pt.level, S((thisTy(prov), pt.body)), nme)(prov)
+      MethodType(pt.level, S((thisTy(prov), pt.body)), nme :: Nil, isInherited = false)(prov)
   }
   
-  private case class MethodDefs(
-    tn: TypeName,
-    parents: List[MethodDefs],
+  /** Represent a set of methods belonging to some owner type.
+    * This includes explicitly declared/defined as well as inherited methods. */
+  private case class MethodSet(
+    ownerType: TypeName,
+    parents: List[MethodSet],
     decls: Map[Str, MethodType],
     defns: Map[Str, MethodType],
   ) {
-    private def &(that: MethodDefs)(tn: TypeName, newdefns: Set[Str])(implicit raise: Raise): MethodDefs =
-      MethodDefs(
+    private def &(that: MethodSet)(tn: TypeName, overridden: Set[Str])(implicit raise: Raise): MethodSet =
+      MethodSet(
         tn,
         this.parents ::: that.parents,
         mergeMap(this.decls, that.decls)(_ & _),
         (this.defns.iterator ++ that.defns.iterator).toSeq.groupMap(_._1)(_._2).flatMap {
           case _ -> Nil => die
           case mn -> (defn :: Nil) => S(mn -> defn)
-          case mn -> defns if newdefns(mn) => N
+          case mn -> defns if overridden(mn) => N  // ignore an inherited method definition if it's overridden
           case mn -> defns =>
             err(msg"An overriding method definition must be given when inheriting from multiple method definitions" -> tn.toLoc
               :: msg"Definitions of method $mn inherited from:" -> N
               :: defns.iterator.map(mt => msg"• ${mt.parents.head}" -> mt.prov.loco).toList)
-            S(mn -> MethodType(defns.head.level, N, tn :: Nil)(
+            S(mn -> MethodType(defns.head.level, N, tn :: Nil, isInherited = false)(
               TypeProvenance(tn.toLoc, "inherited method declaration")))
         }
       )
-    private def addParent(prt: TypeName)(implicit ctx: Ctx): MethodDefs = {
-      val td = ctx.tyDefs(prt.name)
-      def addThis(mt: MethodType): MethodType = mt.copy(body = mt.body.map(b => b.copy(_1 = td.thisTy(mt.prov))))
-      def add(mt: MethodType): MethodType = mt.copy(parents = prt :: mt.parents)
-      copy(decls = decls.view.mapValues(addThis).mapValues(add).toMap, defns = defns.view.mapValues(addThis).toMap)
+    /** When called with top = true, returns a `MethodSet` of the _inherited_ methods only,
+      *   disregarding the current MethodSet's methods... 
+      * Useful for subsumption checking
+      *   as inherited and current methods should be considered separately */
+    def processInheritedMethods(top: Bool)(implicit ctx: Ctx, raise: Raise): MethodSet = {
```

Comment:
Actually the doc of this method is still incomplete. It also performs some checks, possibly raising errors. That's why I renamed the method to include the verb "process". But the doc should mention this more specifically.